src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -79,6 +80,8 @@ public virtual IDisposable Start()
                         contextAccessor.HttpContext = httpContext;
                         await application(httpContext);
                     }
+
+                    return httpContext;
```

Comment:
Is the HttpContext supposed to be usable in OnCompleted? I think the simplest solution is to not support it. It's not like you would be able to change the response at that point.

I'm not a big fan of returning an IDisposable here, because not obvious why it's necessary. It could definitely be a gotcha for anyone implementing or using the IServerFactory interface.

I would even prefer having a "OnTearDown" IHttpResponseFeature that runs after OnCompleted, only because the purpose is more clear.